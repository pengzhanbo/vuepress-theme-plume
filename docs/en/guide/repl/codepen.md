---
title: Code Pen
icon: mingcute:codepen-line
createTime: 2025/10/08 10:41:58
permalink: /en/guide/code/code-pen/
---

The theme supports embedding [CodePen](https://codepen.io/) in Markdown files.

## Configuration

This feature is disabled by default. You can enable it in the configuration file.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      codepen: true, // [!code highlight]
    },
  })
})
```

## Syntax

Basic syntax:

```md
@[codepen](user/slash)
```

Extended options support:

```md
@[codepen preview editable tab="css,result" theme="dark" height="500px" width="100%"](user/slash)
```

- `preview`: Whether to render in preview mode
- `editable`: Whether the content is editable
- `tab`: Default active tabs, defaults to `result`, multiple tabs separated by `,`
- `theme`: Theme, options: `dark` and `light`
- `height`: Container height, defaults to `400px`
- `width`: Container width, defaults to `100%`
- `user`: CodePen username
- `slash`: CodePen pen slug/identifier

## Examples

Input:

```md
@[codepen](leimapapa/RwOZQOW)
```

Output:

@[codepen](leimapapa/RwOZQOW)

**Preview Mode:**

Input:

```md
@[codepen preview](leimapapa/RwOZQOW)
```

Output:

@[codepen preview](leimapapa/RwOZQOW)

**Editable Mode:**

Input:

```md
@[codepen editable tab="html,result"](leimapapa/RwOZQOW)
```

Output:

@[codepen editable tab="html,result"](leimapapa/RwOZQOW)
