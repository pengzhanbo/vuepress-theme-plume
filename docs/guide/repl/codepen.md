---
title: Code Pen
icon: mingcute:codepen-line
createTime: 2024/04/04 10:41:58
permalink: /guide/code/code-pen/
---

主题支持在 Markdown 文件中嵌入 [CodePen](https://codepen.io/)。

## 配置

此功能默认不启用，你可以在配置文件中启用它。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      codepen: true, // [!code highlight]
    },
  })
})
```

## 语法

简单语法：

```md
@[codepen](user/slash)
```

更多选项支持：

```md
@[codepen preview editable tab="css,result" theme="dark" height="500px" width="100%"](user/slash)
```

- `preview`: 是否渲染为预览模式
- `editable`: 是否可编辑
- `tab`: 默认显示的标签， 默认为 `result`，多个使用 `,` 分隔
- `theme`: 主题， 可选值 `dark` 和 `light`
- `height`: 容器高度， 默认为 `400px`
- `width`: 容器宽度， 默认为 `100%`
- `user`: CodePen 用户名
- `slash`: CodePen 代码文件名

## 示例

输入：

```md
@[codepen](leimapapa/RwOZQOW)
```

输出：

@[codepen](leimapapa/RwOZQOW)

**预览模式：**

输入：

```md
@[codepen preview](leimapapa/RwOZQOW)
```

输出：

@[codepen preview](leimapapa/RwOZQOW)

**编辑模式：**

输入：

```md
@[codepen editable tab="html,result"](leimapapa/RwOZQOW)
```

输出：

@[codepen editable tab="html,result"](leimapapa/RwOZQOW)
