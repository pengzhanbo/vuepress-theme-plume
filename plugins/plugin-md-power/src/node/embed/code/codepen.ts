/**
 * @[codepen](user/slash)
 * @[codepen preview](user/slash)
 * @[codepen preview editable title="" height="400px" tab="css,result" theme="dark"](user/slash)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { CodepenTokenMeta } from '../../../shared/index.js'
import { parseRect } from '../../utils/parseRect.js'
import { resolveAttrs } from '../../utils/resolveAttrs.js'
import { createEmbedRuleBlock } from '../createEmbedRuleBlock.js'

const CODEPEN_LINK = 'https://codepen.io/'

export const codepenPlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<CodepenTokenMeta>(md, {
    type: 'codepen',
    syntaxPattern: /^@\[codepen([^\]]*)\]\(([^)]*)\)/,
    meta: ([, info = '', source = '']) => {
      const { attrs } = resolveAttrs(info)
      const [user, slash] = source.split('/')

      return {
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : '400px',
        user,
        slash,
        title: attrs.title,
        preview: attrs.preview,
        editable: attrs.editable,
        tab: attrs.tab ?? 'result',
        theme: attrs.theme,
      }
    },
    content: (meta) => {
      const { title = 'Codepen', height, width } = meta
      const params = new URLSearchParams()

      if (meta.editable) {
        params.set('editable', 'true')
      }
      if (meta.tab) {
        params.set('default-tab', meta.tab)
      }
      if (meta.theme) {
        params.set('theme-id', meta.theme)
      }

      const middle = meta.preview ? '/embed/preview/' : '/embed/'

      const link = `${CODEPEN_LINK}${meta.user}${middle}${meta.slash}?${params.toString()}`
      const style = `width:${width};height:${height};margin:16px auto;border-radius:5px;`

      return `<iframe class="code-pen-iframe-wrapper" src="${link}" title="${title}" style="${style}" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">See the Pen <a href="${CODEPEN_LINK}${meta.user}/pen/${meta.slash}">${title}</a> by ${meta.user} (<a href="${CODEPEN_LINK}${meta.user}">@${meta.user}</a>) on <a href="${CODEPEN_LINK}">CodePen</a>.</iframe>`
    },
  })
}
