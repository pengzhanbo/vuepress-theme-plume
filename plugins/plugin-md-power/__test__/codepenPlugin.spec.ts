import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { codepenPlugin } from '../src/node/embed/code/codepen.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(codepenPlugin)
}

describe('codepenPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[codepen](user/slash)

@[codepen preview](user/slash)

@[codepen preview editable title="codepen" width="100%" height="400px" tab="css,result" theme="dark"](user/slash)
`

    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[codepen]()

@[codepen]xxx

@[codepen preview](
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
