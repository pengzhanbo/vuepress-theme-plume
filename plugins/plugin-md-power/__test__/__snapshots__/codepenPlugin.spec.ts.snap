// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codepenPlugin > should not work 1`] = `
"<iframe class="code-pen-iframe-wrapper" src="https://codepen.io//embed/undefined?default-tab=result" title="Codepen" style="width:100%;height:400px;margin:16px auto;border-radius:5px;" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">See the Pen <a href="https://codepen.io//pen/undefined">Codepen</a> by  (<a href="https://codepen.io/">@</a>) on <a href="https://codepen.io/">CodePen</a>.</iframe><p>@[codepen]xxx</p>
<p>@[codepen preview](</p>
"
`;

exports[`codepenPlugin > should work 1`] = `"<iframe class="code-pen-iframe-wrapper" src="https://codepen.io/user/embed/slash?default-tab=result" title="Codepen" style="width:100%;height:400px;margin:16px auto;border-radius:5px;" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">See the Pen <a href="https://codepen.io/user/pen/slash">Codepen</a> by user (<a href="https://codepen.io/user">@user</a>) on <a href="https://codepen.io/">CodePen</a>.</iframe><iframe class="code-pen-iframe-wrapper" src="https://codepen.io/user/embed/preview/slash?default-tab=result" title="Codepen" style="width:100%;height:400px;margin:16px auto;border-radius:5px;" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">See the Pen <a href="https://codepen.io/user/pen/slash">Codepen</a> by user (<a href="https://codepen.io/user">@user</a>) on <a href="https://codepen.io/">CodePen</a>.</iframe><iframe class="code-pen-iframe-wrapper" src="https://codepen.io/user/embed/preview/slash?editable=true&default-tab=css%2Cresult&theme-id=dark" title="codepen" style="width:100%;height:400px;margin:16px auto;border-radius:5px;" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">See the Pen <a href="https://codepen.io/user/pen/slash">codepen</a> by user (<a href="https://codepen.io/user">@user</a>) on <a href="https://codepen.io/">CodePen</a>.</iframe>"`;
